stan/math/rev/fun/trace_gen_inv_quad_form_ldlt.hpp

```diff
@@ -28,29 +28,141 @@ namespace math {
  * @param B a matrix
  * @return The trace of the inverse quadratic form.
  */
-template <typename EigMat1, typename T2, int R2, int C2, typename EigMat3,
-          require_all_eigen_t<EigMat1, EigMat3>* = nullptr,
-          require_any_st_var<EigMat1, T2, EigMat3>* = nullptr>
-inline var trace_gen_inv_quad_form_ldlt(const EigMat1& D,
-                                        const LDLT_factor<T2, R2, C2>& A,
-                                        const EigMat3& B) {
-  using T3 = value_type_t<EigMat3>;
-  constexpr int R3 = EigMat3::RowsAtCompileTime;
-  constexpr int C3 = EigMat3::ColsAtCompileTime;
+template <typename Td, typename Ta, typename Tb, bool alloc_in_arena,
+          require_all_eigen_t<Td, Tb>* = nullptr,
+          require_any_st_var<Td, Ta, Tb>* = nullptr>
+inline var trace_gen_inv_quad_form_ldlt(
+    const Td& D, const LDLT_factor<Ta, alloc_in_arena>& A, const Tb& B) {
   check_square("trace_gen_inv_quad_form_ldlt", "D", D);
-  check_multiplicable("trace_gen_inv_quad_form_ldlt", "A", A, "B", B);
+  check_multiplicable("trace_gen_inv_quad_form_ldlt", "A", A.matrix(), "B", B);
   check_multiplicable("trace_gen_inv_quad_form_ldlt", "B", B, "D", D);
-  if (D.size() == 0 || A.cols() == 0) {
+
+  if (D.size() == 0 || A.matrix().size() == 0) {
     return 0;
   }
 
-  auto* _impl
-      = new internal::trace_inv_quad_form_ldlt_impl<T2, R2, C2, T3, R3, C3>(
-          D, A, B);
+  if (!is_constant<Ta>::value && !is_constant<Tb>::value
+      && !is_constant<Td>::value) {
+    arena_t<promote_scalar_t<var, Tb>> arena_B = B;
+    arena_t<promote_scalar_t<var, Td>> arena_D = D;
+    auto AsolveB = to_arena(A.ldlt().solve(arena_B.val()));
+    auto BTAsolveB = to_arena(arena_B.val_op().transpose() * AsolveB);
+
+    var res = (arena_D.val() * BTAsolveB).trace();
+
+    reverse_pass_callback(
+        [A, BTAsolveB, AsolveB, arena_B, arena_D, res]() mutable {
+          double C_adj = res.adj();
+
+          forward_as<promote_scalar_t<var, Ta>>(A.matrix()).adj()
+              -= C_adj * AsolveB * arena_D.val_op().transpose()
+                 * AsolveB.transpose();
+          arena_B.adj() += C_adj * AsolveB
+                           * (arena_D.val_op() + arena_D.val_op().transpose());
+          arena_D.adj() += C_adj * BTAsolveB;
+        });
+
+    return res;
+  } else if (!is_constant<Ta>::value && !is_constant<Tb>::value
+             && is_constant<Td>::value) {
+    arena_t<promote_scalar_t<var, Tb>> arena_B = B;
+    arena_t<promote_scalar_t<double, Td>> arena_D = value_of(D);
+    auto AsolveB = to_arena(A.ldlt().solve(arena_B.val()));
```

Comment:
Is the return type here `arena_t<Eigen::MatrixXd>`?

Reply:
Could be a vector or a matrix